### ISSUE TITLE
Using mealpy with LSTM for time series forecasting

### ISSUE THREAD

**Issue by Med-Rokaimi** [thieu1995/mealpy#112]
Hey dear @thieu1995 .
This is really an amazing work which helped me a lot.

However, I have a question please about adapting the example (examples/applications/keras/mha-hyper-parameter-mlp-time-series.py)
to use LSTM instead of MPL.
I have done slight modifications in these parts of the code as follows:
First I prepared my dataset to be:
(batch size, sequence length(time steps), number of features)
Training Shape: (2063, 14, 2) (2063, 1)
Validation  Shape: (259, 14, 2) (259, 1)
Test Shape: (65, 14, 2) (65, 1)

#adding **dropout** for the soultion:
```
   return {

        "batch_size": batch_size,
        "epoch": epoch,
        "opt": opt,
        "learning_rate": learning_rate,
        "network_weight_initial": network_weight_initial,
        "activation": activation,
        "n_hidden_units": n_hidden_units,
        "dropout": dropout,
    }
```
#define the lstm network:
```
def mylstm(structure , data):
    inputs = Input(shape=(data["X_train"].shape[1],data["X_train"].shape[2]))
    lstm_1 = LSTM(structure["n_hidden_units"], return_sequences=False,
                  activation=structure["activation"],
                  dropout=structure["dropout"],
                  kernel_initializer=structure["network_weight_initial"])(inputs)
    #lstm_out=LSTM(structure["n_hidden_units"], return_sequences=False)(lstm_1)
    outputs = Dense(1)(lstm_1)
    model = Model(inputs=inputs, outputs=outputs)
    model.summary()
    return model
```
#add validation for the fit:
```
history = model.fit(

        data["X_train"], data["y_train"], epochs=structure["epoch"], batch_size=structure["batch_size"], verbose=2,
        validation_data=(X_valid, y_valid), shuffle=False,
        ) 
```
finally to call the fitness:
```
model = GWO.OriginalGWO(epoch=20, pop_size=20)
model.solve(problem, termination=term_dict)
```

However, it takes so long just to finalize one epoch. am I doing something wrong in my modification? and kindly, do you think this suits my work for multivariate time series forecasting?

**Comment by thieu1995** [comment]
Hi @Med-Rokaimi ,

You are tuning hyper-parameter of deep learning model. That is why it takes time. In each epoch it will run (epoch*pop_size) = 400 LSTM networks. Each network takes around 2 minutes to train then 400 nets are about 13 hours to train a single epoch. 
It is just like GridSearchCV in scikit-learn. However, GridSearchCV is brute force method. Here, metaheuristics won't try all of possible combinations of pameters. It will faster than GridSearchCV.

**Comment by Med-Rokaimi** [comment]
Thank you very much for your help. so clear.

**Comment by thieu1995** [comment]
@Med-Rokaimi ,

There is no best optimizer for all the problems (No-free launch theory). So you can try algorithms as much as possible to compare them. For example, SHADE, LSHADE, AVOA, WOA, SSA, TLO, AAEO, RUN, INFO... just try them.
This is big problem, so should increase epoch and decrease pop_size. For example, epoch=100, pop_size=20 (or 30). Why?
Because pop_size doesn't affect much comparing to epoch. Also set pop size bigger then more processing time on each epoch.

**Comment by Med-Rokaimi** [comment]
Thank you for this help. so useful. 
May I ask about fine tuning other parameter's related to the preparation dataset it self. For example, I want to optimize the sequence length (look-back) in the time series problem. This process normally carried out before the dataset split. 
Thank you

**Comment by thieu1995** [comment]
Hi @Med-Rokaimi ,

It is really the same process of building LSTM like above. Now you have to design a function that can preprocess data, build LSTM, train LSTM, and return loss (RMSE or MSE,...) as fitness value. 

Before, you preprocess the data at Global Scope of Python, now you just have to move it to fitness function.

**Comment by thieu1995** [comment]
I will close the issue now since there is no more question.
